AEROGEAR-9184

laurafitzgerald opened the pull request:
**JIRA**
https://issues.jboss.org/browse/AEROGEAR-9184

**Tasks**
- [x] Add some dependencies in order to ad the new api for the a MobileSecurityApp
- [x] Generate the stuff
- [x] Have the operator respond to the creation of the app cr, aka create the required SDK config
- [x] Put the right stuff in the Confg Map
- [x] Edit - Have the config map be created if the cr is in a namespace which is provided by the APP_NAMESPACE variable.

**Verification Steps**

`make cluster/clean`
`make cluster/prepare`
`make install`

Run the following to verify the _correct_ namespace:
`make example-app/apply`
Verify that the sdk config map is created in the `app-metrics-apps` namespace with the name `<cr-app-name>-metrics`
Verify the contents of the config map is `url` : `the route for the metrics service`

Run the following to verify the _incorrect_ namespace:
`make example-app/delete`
Change the APP_NAMESPACES variable in the Makefile to `app-metrics`
`make example-app/apply`

Verify the cr is created in the `app-metrics` ns
Verify that the sdk config map is **not** created in the `app-metrics` namespace
Verify that the operator produces a log with the following content `"app-metrics/example-app","error":"The app cr example-app was created in a namespace which is not present in the APP_NAMESPACES environment variable provided to the operator","`

laurafitzgerald committed:
adding code-gen dependencies

laurafitzgerald committed:
adding generated files for the app cr

laurafitzgerald committed:
creates the config map in a specified namespace

laurafitzgerald commented on deploy/operator.yaml:
```diff
@@ -16,15 +16,16 @@ spec:
         - name: app-metrics-operator
           # Replace this with the built image name
-          image: REPLACE_IMAGE
+          image: quay.io/aerogear/app-metrics-operator:ag-9184
```
This is pushed for convenience. The dev image is built and pushed for testing. I'll remove it before merging this one.

grdryn replied:
If you don't want to have to do this for PRs, you can just put this in the verification steps:

```
sed -e "s|quay.io/aerogear/app-metrics-operator:master|quay.io/aerogear/app-metrics-operator:ag-9184|g" deploy/operator.yaml | kubectl create -f -
```

Or if you wanted a make task, something like this could go in there:
```
sed -e "s|quay.io/aerogear/app-metrics-operator:master|quay.io/aerogear/app-metrics-operator:$(git rev-parse --short HEAD)|g" deploy/operator.yaml | kubectl create -f -
```

laurafitzgerald commented on deploy/role.yaml:
```diff
@@ -14,6 +14,13 @@ rules:
   - events
```
This is a copy of what's in https://github.com/aerogear/app-metrics-operator/pull/2

aliok commented on deploy/crds/metrics_v1alpha1_appmetricsapp_cr.yaml:
```diff
@@ -0,0 +1,7 @@
+apiVersion: metrics.aerogear.com/v1alpha1
+kind: AppMetricsApp
```
The kind looks a bit strange :)

Was it a mistake? Or if on purpose, can you explain what should resource do please?

aliok replied:
Ah, this would represent a client app...
Maybe AppMetricsClient is a better name?
Or AppMetricsSource?

laurafitzgerald replied:
I'm basing it on the MSS consumption cr kind which is MobileSecurityServiceApp, aka ServiceNameApp. It just happens that the name of this service has App in it already. Not sure I like it either. Thoughts on an alternative? In ups we use `variant` because that's what's created in the service instead of `app` which is what's created in the MSS service. There's no service here where something is created because of the bind. All we need to do is create the config map for use by MDC in mobile-services.json.

grdryn replied:
Yeah, I'm not sure about the name of the installation CR either (AppMetricsService), happy to change that to if it helps.

@laurafitzgerald also note that in #2 I've changed the apiVersion to use .org instead of .com, to match with the other operators (I accidentally had it as .com when I created the project).

grdryn replied:
How about `AppMetricsConfig`? :thinking:

aliok replied:
sounds better than AppMetricsApp :) +1 on that

laurafitzgerald replied:
+1 Going with `AppMetricsConfig`, will push the change soon.

aliok commented:
@laurafitzgerald Not related to your PR, but I noticed these env var names are stolen badly :)

https://github.com/aerogear/app-metrics-operator/blob/55158c28ad2380e16e94ba1a1aca29ea05983cff/pkg/config/config.go#L27-L28

Can you also fix that?

laurafitzgerald commented:
@aliok sure, I'll fix them in a seperate pr.

grdryn commented:
Oops, sorry about that :see_no_evil:

laurafitzgerald commented:
@grdryn @aliok this is now ready for final review and verification. The steps are updated above.

laurafitzgerald commented on pkg/controller/appmetricsapp/appmetricsapp_controller.go:
```diff
@@ -0,0 +1,204 @@
+	}
+}
+
+func GetAppNamespaces() (string, error) {
```
The two functions here are lifted from MSS.

grdryn replied:
I'd change them both to an initial lowercase letter, since they don't need to be exported from this package.

laurafitzgerald replied:
Done

grdryn commented on Gopkg.lock:
```diff
@@ -1071,9 +1114,12 @@
   analyzer-version = 1
   input-imports = [
     "github.com/go-openapi/spec",
+<<<<<<< HEAD
```
this looks like a merge/rebase conflict marker

laurafitzgerald replied:
Done

grdryn commented on Makefile:
```diff
@@ -1,7 +1,14 @@
+
+QUAY_ORG=aerogear
+QUAY_IMAGE=app-metrics-operator
+DEV_TAG=ag-9184
```
Could this be generated instead of hardcoded? If it's hardcoded we have to change it every time. I tend to use `git rev-parse --short HEAD`

laurafitzgerald replied:
Yes, that shouldn't have been committed. the dev tag can be `dev`.

laurafitzgerald replied:
Done

grdryn commented on pkg/apis/metrics/v1alpha1/appmetricsapp_types.go:
```diff
@@ -0,0 +1,45 @@
+type AppMetricsAppSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
+	// Add custom validation using kubebuilder tags: https://book.kubebuilder.io/beyond_basics/generating_crd.html
```
Could you remove these boiler-plate comments? If you add a custom comment, it should get added to the CRD I think.

laurafitzgerald replied:
Done

grdryn commented on pkg/apis/metrics/v1alpha1/appmetricsapp_types.go:
```diff
@@ -0,0 +1,45 @@
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
+	// Add custom validation using kubebuilder tags: https://book.kubebuilder.io/beyond_basics/generating_crd.html
+	Name string `json:"app-name"`
```
Is there an advantage to having a `spec["app-name"]` field, when there's the `metadata.name` field that must be specified anyway? If someone changes this, the name of the CR can't be changed, so they'd be out of sync.

If it's necessary, I'd suggest changing the json tag to be `appName`, since that's more normal/consistent in resource definitions (it can also be icky in some programming languages to reference fields by name if the name has a hyphen in it E.g. `spec["app-name"]` is needed instead of `spec.appName` in e.g. JavaScript, and it's not outside the realm of possibility that the MDC would read this in some future version)

laurafitzgerald replied:
Yes, we had a separate appName defined in the spec, the reason for this was that it could be updated in the MDC, which (I believe) metadata.name is not updateable. There was a discussion about removing the ability to update the app name in the MDC which I believe we are doing now so that may remove the need for this additional spec. But we should check that with @ziccardi or @craicoverflow before removing it. For now I will update to the better syntax until we can get some guidance from the guys.

ziccardi replied:
@grdryn @laurafitzgerald Correct. We already removed the ability to update the app name in the current master.

laurafitzgerald replied:
So we should not use spec.appName?

laurafitzgerald replied:
@ziccardi

laurafitzgerald replied:
app-name removed. Done

grdryn commented on pkg/controller/appmetricsapp/appmetricsapp_controller.go:
```diff
@@ -0,0 +1,204 @@
+		return err
+	}
+
+	// TODO(user): Modify this to be the types you create that are owned by the primary resource
```
this todo comment can be removed

laurafitzgerald replied:
Done

grdryn commented on pkg/controller/appmetricsapp/appmetricsapp_controller.go:
```diff
@@ -0,0 +1,204 @@
+	return ns, nil
+}
+
+// IsValidOperatorNamespace return true when the namespace informed is declared in the ENV VAR APP_NAMESPACES
```
This comment could do with some love: it has the function name wrong, and I don't know what it means by 'informed'

laurafitzgerald replied:
👍 copy pasta. Will improve.

laurafitzgerald replied:
Done

grdryn commented on pkg/controller/appmetricsapp/appmetricsapp_controller.go:
```diff
@@ -0,0 +1,204 @@
+
+		return false, err
+	}
+	for _, ns := range strings.Split(appNamespacesEnvVar, ";") {
```
The fact that this string gets split on semi-colon rather than comma or space or comma-space, was not obvious to me until now. :crying_cat_face: Maybe there should be a yaml comment in the `operator.yaml` (and README)? I'd suggest changing it to comma, but that probably makes it inconsistent with the other operator you took it from, right?

laurafitzgerald replied:
Yes, I used the semi-colon as it was used elsewhere. I prefer a comma also myself. I'm also unsure about the use of a list as our use case is for one namespace. Perhaps there are users upstream who would prefer to be able to provide a list but it hasn't been explicitly asked for yet. Interested in your thoughts also there.

laurafitzgerald replied:
Changed to comma. Done

grdryn commented on pkg/controller/appmetricsapp/appmetricsapp_controller.go:
```diff
@@ -0,0 +1,204 @@
+		return reconcile.Result{}, err
+	}
+
+	if !is {
```
the `IsValidAppNamespace` only ever seems to return `true, nil` or `false, err`, which makes the first boolean return value (which we clearly had trouble naming :sweat_smile:) and this check redundant I think (in the current implementation of that function, this block will never be entered if execution gets this far).

laurafitzgerald replied:
Ok. I'll recheck this. It's a copy pasta so I just took it as is. Happy to make improvements. :)

laurafitzgerald replied:
Done

grdryn commented on pkg/controller/appmetricsapp/appmetricsapp_controller.go:
```diff
@@ -0,0 +1,204 @@
+	listOptions.InNamespace("app-metrics")
+	listOptions.MatchingLabels(map[string]string{"component": "aerogear-app-metrics"})
+	err = r.client.List(context.TODO(), listOptions, routeList)
+
```
I'd put this empty line above the previous line (assign to `err` and check its value in the same visual block). 

Actually I think that line 115 above could even be part of the `if` construct on line 117, since the only thing returned there is the `err` that will be checked and not used outside that check.

laurafitzgerald replied:
Done

grdryn commented on pkg/controller/appmetricsapp/appmetricsapp_controller.go:
```diff
@@ -0,0 +1,204 @@
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      configmapname,
+			Namespace: "mobile-developer-console",
```
I think this should be `cr.Namespace` rather than a hard-coded assumption (we only want to put the config map into a namespace called "mobile-developer-console" if that's where a CR was created, right?).

laurafitzgerald replied:
yes.

laurafitzgerald replied:
Done

grdryn commented on pkg/controller/appmetricsapp/appmetricsapp_controller.go:
```diff
@@ -0,0 +1,204 @@
+	}
+}
+
+func GetAppNamespaces() (string, error) {
```
This is an oddly specific version of a more generic function that could have the signature `func getEnvVar(envVar string) (string, error)`, which may or may not exist. My guess is that it doesn't exist because it's not really necessary.

laurafitzgerald replied:
Done

grdryn commented on pkg/controller/appmetricsapp/appmetricsapp_controller.go:
```diff
@@ -0,0 +1,204 @@
+// IsValidOperatorNamespace return true when the namespace informed is declared in the ENV VAR APP_NAMESPACES
+func IsValidAppNamespace(namespace string) (bool, error) {
+	appNamespacesEnvVar, err := GetAppNamespaces()
+	if err != nil {
```
These two lines could be replaced with the following (removing the need for the `GetAppNamespaces()` function at all):

```go
appNamespacesEnvVar, found := os.LookupEnv("APP_NAMESPACES")
if !found {
```

laurafitzgerald replied:
Done

grdryn commented on pkg/controller/appmetricsapp/appmetricsapp_controller.go:
```diff
@@ -0,0 +1,204 @@
+
+	routeList := &routev1.RouteList{}
+	listOptions := &client.ListOptions{}
+	listOptions.InNamespace("app-metrics")
```
Rather than hard-coding `"app-metrics"` here, I'd use `k8sutil.GetOperatorNamespace()` (not sure what that function does for `operator-sdk up local` though).

laurafitzgerald replied:
Done

grdryn commented on Makefile:
```diff
@@ -1,7 +1,14 @@
 NAMESPACE=app-metrics
+APP_NAMESPACES=app-metrics-apps
```
this is different to the default in the operator.yaml, not sure if that's important.

laurafitzgerald replied:
You are correct. :) It's done this way so that the operator can be decoupled from MDC in the community if someone wishes to do that. I think I need to update the value passed to the operator env var, though 🤔

grdryn commented on pkg/controller/appmetricsapp/appmetricsapp_controller.go:
```diff
@@ -0,0 +1,204 @@
+			Namespace: "mobile-developer-console",
+		},
+		Data: map[string]string{
+			"SDKConfig": fmt.Sprintf("{\"url\": \"https://%s\"}", host),
```
This is fine for now, but if we add more than this single field to the JSON object, I'd consider using something from the json package rather than hand-crafting it and escaping all of those quotes.

grdryn reviewed (dismissed):
In case it's hidden (because that file doesn't show by default), there are conflict markers in Gopkg.lock in this PR that need to be removed. Further comments inline, but generally good and in the right direction! :+1:

laurafitzgerald commented:
@grdryn the two most recent commits are based on your feedback items. Could you take a look when you have some time?

aliok commented:
Gave this PR a try locally.

I think it all works correctly, but there's some incompatibility with what's written in the verification instructions. Maybe they're outdated?

grdryn commented on deploy/operator.yaml:
```diff
@@ -15,15 +15,17 @@ spec:
-                  fieldPath: metadata.namespace
+              value: ""
+            - name: APP_NAMESPACES
+              # Replace this with a ; seperated list of namespaces which the operator will action app crs
```
```suggestion
              # Replace this with a ; separated list of namespaces in which the operator will act on app crs
```

laurafitzgerald replied:
Done

grdryn commented on pkg/apis/metrics/v1alpha1/appmetricsapp_types.go:
```diff
@@ -0,0 +1,42 @@
+// AppMetricsAppSpec defines the desired state of AppMetricsApp
+// +k8s:openapi-gen=true
+type AppMetricsAppSpec struct {
+	Name string `json:"appName"`
```
It sounds like based on the conversation in the previous review, that this can be removed. Wdyt?

laurafitzgerald replied:
Removed. 👍

laurafitzgerald committed:
appmetricsapp -> appmetricsconfig

laurafitzgerald committed:
list comma and remove app-name

laurafitzgerald commented:
@aliok yes those were out of date. They are up to date now and I've pushed additional changes as suggested.  All the current feedback is addressed so this is ready to be verified now. Thanks for all the feedback. :)

grdryn reviewed (approved):
All looks good to me: reviewed and verified.

One thing from the verification steps is that the log message was different than in the steps, but I don't think that's a big deal:

```
{"level":"error","ts":1562689049.299172,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"appmetricsconfig-controller","request":"app-metrics/example-app","error":"The app cr example-app was created in a namespace which is not present in the APP_NAMESPACES environment variable provided to the operator","stacktrace":"github.com/aerogear/app-metrics-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:217\ngithub.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
```

Additionally, in the code, the `DEV_TAG` stuff still exists and/or is specific to this PR.

laurafitzgerald commented:
> All looks good to me: reviewed and verified.
> 
> One thing from the verification steps is that the log message was different than in the steps, but I don't think that's a big deal:
> 
> ```
> {"level":"error","ts":1562689049.299172,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"appmetricsconfig-controller","request":"app-metrics/example-app","error":"The app cr example-app was created in a namespace which is not present in the APP_NAMESPACES environment variable provided to the operator","stacktrace":"github.com/aerogear/app-metrics-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:217\ngithub.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
> ```
> 
> Additionally, in the code, the `DEV_TAG` stuff still exists and/or is specific to this PR.

@grdryn thanks for the approval. I will remove the DEV_TAG stuff and I've updated the description for the verification.  I changed the log message when I made the improvements.

laurafitzgerald commented:
I've cleaned up the commits and reverified everything works my end. Merging as it has approval.
